test/mix/tasks/gradient_test.exs

```diff
@@ -219,7 +237,283 @@ defmodule Mix.Tasks.GradientTest do
+    end
+
+    @tag umbrella: %{enabled: true}, app_a: %{no_config: true}, app_b: %{no_config: true}
+    test "when gradient is enabled for umbrella, checks all files" do
+      assert %{
+               "app_a" => [
+                 "_build/dev/lib/app_a/ebin/Elixir.AppA.beam",
+                 "_build/dev/lib/app_a/ebin/Elixir.AppAHelper.beam"
+               ],
+               "app_b" => [
+                 "_build/dev/lib/app_b/ebin/Elixir.AppB.beam",
+                 "_build/dev/lib/app_b/ebin/Elixir.AppBHelper.beam"
+               ]
+             } == run_task_and_return_files()
+    end
+
+    @tag umbrella: %{enabled: false}, app_a: %{no_config: true}, app_b: %{no_config: true}
+    test "when gradient is disabled for umbrella, doesn't check any files" do
+      assert %{} == run_task_and_return_files()
+    end
+
+    @tag umbrella: %{enabled: false}, app_a: %{enabled: true}
+    test "gradient can be enabled for a subapp even if disabled for umbrella" do
+      assert %{
+               "app_a" => [
+                 "_build/dev/lib/app_a/ebin/Elixir.AppA.beam",
+                 "_build/dev/lib/app_a/ebin/Elixir.AppAHelper.beam"
+               ]
+             } == run_task_and_return_files()
+    end
+
+    @tag umbrella: %{enabled: true}, app_a: %{enabled: false}
+    test "gradient can be enabled for the umbrella and disabled for a subapp" do
+      assert %{
+               "app_b" => [
+                 "_build/dev/lib/app_b/ebin/Elixir.AppB.beam",
+                 "_build/dev/lib/app_b/ebin/Elixir.AppBHelper.beam"
+               ]
+             } == run_task_and_return_files()
+    end
+
+    @tag umbrella: %{enabled: true, overrides: true},
+         edit_files: %{"app_a/lib/app_a_helper.ex" => false}
+    test "individual files can be disabled" do
+      assert %{
+               "app_a" => [
+                 "_build/dev/lib/app_a/ebin/Elixir.AppA.beam"
+               ],
+               "app_b" => [
+                 "_build/dev/lib/app_b/ebin/Elixir.AppB.beam",
+                 "_build/dev/lib/app_b/ebin/Elixir.AppBHelper.beam"
+               ]
+             } == run_task_and_return_files()
+    end
+
+    @tag umbrella: %{enabled: false, overrides: true}, edit_files: %{"app_a/lib/app_a.ex" => true}
+    test "individual files can be enabled" do
+      assert %{"app_a" => ["_build/dev/lib/app_a/ebin/Elixir.AppA.beam"]} ==
+               run_task_and_return_files()
+    end
+  end
+
+  # Run the task in the current process. Useful for running on a single file.
   def run_task(args), do: capture_io(fn -> Mix.Tasks.Gradient.run(args) end)
 
-  def test_opts(opts), do: ["--no-comile", "--no-deps"] ++ opts
+  # Run the task in a new process, via the shell. Useful for running on an entire project.
+  def run_shell_task(dir, args) do
+    previous_cwd = File.cwd!()
+
+    try do
+      # cd to the specified dir
+      File.cd!(dir)
+
+      # mix clean first
+      assert {_, 0} = System.cmd("mix", ["clean"])
```

Comment:
Added a `mix clean` before any tests for `simple_umbrella_app`, in order to reproduce the bug I mentioned.